### PR TITLE
Refactor catalog serialize/deserialize

### PR DESCRIPTION
**Pull request by ray6080**

### CODE REVIEW

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -312,7 +200,7 @@ void CatalogContent::addScalarMacroFunction(
     macros.emplace(std::move(name), std::move(macro));
 }
 
-void CatalogContent::validateStorageVersion(storage_version_t savedStorageVersion) const {
+void CatalogContent::validateStorageVersion(storage_version_t savedStorageVersion) {
```
Comment (acquamarin) [inline src/catalog/catalog.cpp:203]: I don't think this function changes any fields of catalogContent, why not make it const?

File: src/catalog/catalog.cpp [hunk 2]
```diff
@@ -257,24 +147,22 @@ void CatalogContent::readFromFile(const std::string& directory, DBFileType dbFil
...
     SerDeser::deserializeValue<uint64_t>(numNodeTables, fileInfo.get(), offset);
     SerDeser::deserializeValue<uint64_t>(numRelTables, fileInfo.get(), offset);
     table_id_t tableID;
     for (auto i = 0u; i < numNodeTables; i++) {
-        SerDeser::deserializeValue<uint64_t>(tableID, fileInfo.get(), offset);
-        nodeTableSchemas[tableID] = std::make_unique<NodeTableSchema>();
-        SerDeser::deserializeValue<NodeTableSchema>(
-            *nodeTableSchemas[tableID], fileInfo.get(), offset);
+        SerDeser::deserializeValue<table_id_t>(tableID, fileInfo.get(), offset);
+        nodeTableSchemas[tableID] = ku_static_unique_pointer_cast<TableSchema, NodeTableSchema>(
```
Comment (acquamarin) [inline src/catalog/catalog.cpp:159]: I think you don't need to do a pointer cast, c++ will do it automatically for you

File: src/catalog/table_schema.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    case RelMultiplicity::ONE_MANY: {
+        return "ONE_MANY";
+    }
+    default:
+        throw CatalogException("Cannot convert rel multiplicity to std::string.");
+    }
+}
+
+void Property::serialize(FileInfo* fileInfo, uint64_t& offset) const {
+    SerDeser::serializeValue<std::string>(name, fileInfo, offset);
```
Comment (acquamarin) [inline src/catalog/table_schema.cpp:45]: Actually, you don't need to write the specialization type here. C++ will auto infer the correct function based on the type of name

File: src/catalog/table_schema.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    throw InternalException(
+        StringUtils::string_format("Property with id={} not found.", propertyID));
+}
+
+void TableSchema::serialize(FileInfo* fileInfo, uint64_t& offset) {
+    SerDeser::serializeValue<std::string>(tableName, fileInfo, offset);
+    SerDeser::serializeValue<table_id_t>(tableID, fileInfo, offset);
+    SerDeser::serializeValue<TableType>(tableType, fileInfo, offset);
+    SerDeser::serializeValue<uint64_t>(properties.size(), fileInfo, offset);
+    for (auto& property : properties) {
```
Comment (acquamarin) [inline src/catalog/table_schema.cpp:113]: You can reuse the serializeVector interface in ser/deser

File: src/include/catalog/catalog_content.h [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    std::unordered_map<std::string, common::table_id_t> relTableNameToIDMap;
+    common::table_id_t nextTableID;
+    std::unique_ptr<function::BuiltInVectorFunctions> builtInVectorFunctions;
+    std::unique_ptr<function::BuiltInAggregateFunctions> builtInAggregateFunctions;
+    std::unique_ptr<function::BuiltInTableFunctions> builtInTableFunctions;
+    std::unordered_map<std::string, std::unique_ptr<function::ScalarMacroFunction>> macros;
+};
+
+} // namespace catalog
+} // namespace kuzu
```
Comment (acquamarin) [inline src/include/catalog/catalog_content.h:166]: newline

File: src/include/catalog/table_schema.h [hunk 6]
```diff
@@ -77,31 +82,45 @@ struct TableSchema {
...
-        : TableSchema{std::move(tableName), tableID, true /* isNodeTable */, std::move(properties)},
+        : TableSchema{std::move(tableName), tableID, TableType::NODE, std::move(properties)},
           primaryKeyPropertyID{primaryPropertyId} {}
 
     inline void addFwdRelTableID(common::table_id_t tableID) { fwdRelTableIDSet.insert(tableID); }
     inline void addBwdRelTableID(common::table_id_t tableID) { bwdRelTableIDSet.insert(tableID); }
 
     inline Property getPrimaryKey() const { return properties[primaryKeyPropertyID]; }
 
+    void serializeInternal(common::FileInfo* fileInfo, uint64_t& offset) final;
```
Comment (acquamarin) [inline src/include/catalog/table_schema.h:119]: make `serializeinternal` private.

**Review by acquamarin — APPROVED** [review]

File: src/catalog/catalog.cpp (hunk 1 above)
Comment (ray6080) [reply src/catalog/catalog.cpp:203]: This function is changed to static. I don't think it requires any fields of the class.

File: src/catalog/catalog.cpp (hunk 2 above)
Comment (ray6080) [reply src/catalog/catalog.cpp:159]: This is a downward casting, which cannot be handled automatically by cpp.

File: src/catalog/table_schema.cpp (hunk 4 above)
Comment (ray6080) [reply src/catalog/table_schema.cpp:113]: Added `serializeVectorOfObjects` and `deserializeVectorOfObjects` in ser/deser for now.

**Pull request merged**